calour/dsfdr.py

```diff
@@ -114,106 +123,115 @@ def dsfdr(data, labels, method='meandiff', transform='rankdata', alpha=0.1, nump
         number of permutations to perform
 
+    fdrmethod : str
+        the FDR procedure to determine significant bacteria
+        'dsfdr' : discrete FDR method
+        'bhfdr' : Benjamini-Hochberg FDR method
+        'byfdr' : Benjamini-Yekutielli FDR method 
+        'filterBH' : Benjamini-Hochberg FDR method with filtering
+
     output:
     reject : np array of bool (length N)
         True for OTUs where the null hypothesis is rejected
     tstat : np array of float (length N)
-        the t-statistic value for each OTU (for effect size)
+        the test statistic value for each OTU (for effect size)
     pvals : np array of float (length N)
-        the p-value for each OTU
+        the p-value for each OTU    
     '''
+    
     logger.debug('dsfdr using fdr method: %s' % fdrmethod)
-    data = data.copy()
+
+    data=data.copy()
+
+    if fdrmethod == "filterBH":
+        index=[]
+        n0 = np.sum(labels==0)
+        n1 = np.sum(labels==1)
+
+        for i in range(np.shape(data)[0]):
+            nonzeros = np.count_nonzero(data[i,:])
+            if nonzeros < min(n0,n1):
+                pval_min = (comb(n0, nonzeros, exact=True) + comb(n1, nonzeros, exact=True))/comb(n0+n1, nonzeros)
+                if pval_min <= alpha:
+                    index.append(i)  
+            else:
+                index.append(i)
+        data = data[index,:]    
+
+
     # transform the data
     if transform == 'rankdata':
-        data = rankdata(data)
-    elif transform == 'log2data':
-        data = log2data(data)
-    elif transform == 'binarydata':
-        data = binarydata(data)
-    elif transform == 'normdata':
-        data = normdata(data)
-
-    numbact = np.shape(data)[0]
+           data = rankdata(data)
+     elif transform == 'log2data':
+         data = log2data(data)
+     elif transform == 'binarydata':
+         data = binarydata(data)
+     elif transform == 'normdata':
+         data = normdata(data)
+ 
+     numbact = np.shape(data)[0]
+ 
+     labels = labels.copy()
+
+    numbact=np.shape(data)[0]
+    labels=labels.copy()
 
-    labels = labels.copy()
     if method == "meandiff":
-        # do matrix multiplication based calculation
+        # fast matrix multiplication based calculation
         method = meandiff
-        tstat = method(data, labels)
-        t = np.abs(tstat)
-        numsamples = np.shape(data)[1]
-        p = np.zeros([numsamples, numperm])
-        k1 = 1 / np.sum(labels == 0)
-        k2 = 1 / np.sum(labels == 1)
+         tstat = method(data, labels)
+         t = np.abs(tstat)
+         numsamples = np.shape(data)[1]
+         p = np.zeros([numsamples, numperm])
+         k1 = 1 / np.sum(labels == 0)
+         k2 = 1 / np.sum(labels == 1)
         for cperm in range(numperm):
             np.random.shuffle(labels)
             p[labels == 0, cperm] = k1
-        p2 = np.ones(p.shape) * k2
-        p2[p > 0] = 0
+         p2 = np.ones(p.shape) * k2
+         p2[p > 0] = 0
```

Comment:
also lines 198-199 have an additional space before the commands, which is confusing for the unit test (and for reading the code :) )